java/vector/src/test/java/org/apache/arrow/vector/TestBitVectorHelper.java

```diff
@@ -149,4 +149,87 @@ public void testAllBitsSet() {
       assertFalse(BitVectorHelper.checkAllBitsEqualTo(validityBuffer, bitLength, true));
     }
   }
+
+  @Test
+  public void testConcatBits() {
+    try (RootAllocator allocator = new RootAllocator(1024 * 1024)) {
+      try (ArrowBuf buf1 = allocator.buffer(1024);
+           ArrowBuf buf2 = allocator.buffer(1024);
+           ArrowBuf output = allocator.buffer(1024)) {
+
+        buf1.setZero(0, buf1.capacity());
+        buf2.setZero(0, buf2.capacity());
+
+        final int maxCount = 100;
+        for (int i = 0; i < maxCount; i++) {
+          if (i % 3 == 0) {
+            BitVectorHelper.setValidityBitToOne(buf1, i);
+            BitVectorHelper.setValidityBitToOne(buf2, i);
+          }
+        }
+
+        // test the case where the number of bits for both sets are multiples of 8.
+        concatAndVerify(buf1, 40, buf2, 48, output);
+
+        // only the number of bits in the first set is a multiple of 8
+        concatAndVerify(buf1, 32, buf2, 47, output);
+
+        // only the number of bits in the second set is a multiple of 8
+        concatAndVerify(buf1, 31, buf2, 48, output);
+
+        // neither set has a size that is a multiple of 8
+        concatAndVerify(buf1, 27, buf2, 52, output);
+
+        // the remaining bits in the second set is spread in two bytes
+        concatAndVerify(buf1, 31, buf2, 55, output);
+      }
+    }
+  }
+
+  @Test
+  public void testConcatBitsInPlace() {
+    try (RootAllocator allocator = new RootAllocator(1024 * 1024)) {
+      try (ArrowBuf buf1 = allocator.buffer(1024);
+           ArrowBuf buf2 = allocator.buffer(1024)) {
+
+        buf1.setZero(0, buf1.capacity());
+        buf2.setZero(0, buf2.capacity());
+
+        final int maxCount = 100;
+        for (int i = 0; i < maxCount; i++) {
+          if (i % 3 == 0) {
+            BitVectorHelper.setValidityBitToOne(buf1, i);
+            BitVectorHelper.setValidityBitToOne(buf2, i);
+          }
+        }
+
+        // test the case where the number of bits for both sets are multiples of 8.
+        concatAndVerify(buf1, 40, buf2, 48, buf1);
+
+        // only the number of bits in the first set is a multiple of 8
+        concatAndVerify(buf1, 32, buf2, 47, buf1);
+
+        // only the number of bits in the second set is a multiple of 8
+        concatAndVerify(buf1, 31, buf2, 48, buf1);
+
+        // neither set has a size that is a multiple of 8
+        concatAndVerify(buf1, 27, buf2, 52, buf1);
+
+        // the remaining bits in the second set is spread in two bytes
+        concatAndVerify(buf1, 31, buf2, 55, buf1);
+      }
+    }
+  }
+
+  private void concatAndVerify(ArrowBuf buf1, int count1, ArrowBuf buf2, int count2, ArrowBuf output) {
+    BitVectorHelper.concatBits(buf1, count1, buf2, count2, output);
+    for (int i = 0; i < count1 + count2; i++) {
+      int result = BitVectorHelper.get(output, i);
+      if (i < count1) {
```

Comment:
I think it would be clearer and less reliant on the input setup to write this as:

outputIdx =0;
for (int i = 0; i < count1; i++, outputIdx++) {
   assertEquals(BitVectorHelper.get(output, outputIdx), BitVectorHelper.get(buf1, i))
}
for (int i = 0; i < count2; i++, outputIdx++) {
   assertEquals(BitVectorHelper.get(output, outputIdx), BitVectorHelper.get(buf2, i))
}

with this setup, for instance you could "fuzz" by generating random bit strings in addition to the set pattern here.

Reply:
Good suggestio. Revised accordingly. Thank you.